node/pkg/driver/device_connectivity/device_connectivity_fc.go

```diff
@@ -0,0 +1,324 @@
+			}
+
+		}
+	}
+
+	var mps string
+	for key := range devicePathTosysfs {
+		mps += ", " + key
+	}
+	logger.Infof("GetMpathDevice: Found multipath devices: [%s] that relats to lunId=%d and arrayIdentifiers=%s", mps, lunId, arrayIdentifiers)
+
+	if len(devicePathTosysfs) > 1 {
+		return "", &MultipleDmDevicesError{volumeId, lunId, arrayIdentifiers, devicePathTosysfs}
+	}
+
+	var md string
+	for md = range devicePathTosysfs {
+		break // because its a single value in the map(1 mpath device, if not it should fail above), so just take the first
+	}
+	return md, nil
+}
+
+func (r OsDeviceConnectivityFc) FlushMultipathDevice(mpathDevice string) error {
+	return r.HelperScsiGeneric.FlushMultipathDevice(mpathDevice)
+}
+
+func (r OsDeviceConnectivityFc) RemovePhysicalDevice(sysDevices []string) error {
+	return r.HelperScsiGeneric.RemovePhysicalDevice(sysDevices)
+}
+
+// ============== OsDeviceConnectivityHelperFcInterface ==========================
+
+//go:generate mockgen -destination=../../../mocks/mock_OsDeviceConnectivityHelperFcInterface.go -package=mocks github.com/ibm/ibm-block-csi-driver/node/pkg/driver/device_connectivity OsDeviceConnectivityHelperFcInterface
+
+type OsDeviceConnectivityHelperFcInterface interface {
+	/*
+		This is helper interface for OsDeviceConnectivityFc.
+		Mainly for writting clean unit testing, so we can Mock this interface in order to unit test OsDeviceConnectivityFc logic.
+	*/
+	WaitForPathToExist(devicePath string, maxRetries int, intervalSeconds int) ([]string, bool, error)
+	GetMultipathDisk(path string) (string, error)
+	GetFcHostIDs() ([]int, error)
+}
+
+type OsDeviceConnectivityHelperFc struct {
+	executer executer.ExecuterInterface
+}
+
+const (
+	FC_HOST_SYSFS_PATH = "/sys/class/fc_host"
+)
+
+func NewOsDeviceConnectivityHelperFc(executer executer.ExecuterInterface) OsDeviceConnectivityHelperFcInterface {
+	return &OsDeviceConnectivityHelperFc{executer: executer}
+}
+
+func (o OsDeviceConnectivityHelperFc) GetFcHostIDs() ([]int, error) {
+	/*
+		Description:
+			This function find all the hosts IDs under directory /sys/class/fc_host/"
+			So the function goes over all the above hosts and return back only the host numbers as a list.
+	*/
+
+	targetNamePath := FC_HOST_SYSFS_PATH + "/host*/port_state"
+	var HostIDs []int
+	matches, err := o.executer.FilepathGlob(targetNamePath)
+	if err != nil {
+		logger.Errorf("Error while Glob targetNamePath : {%v}. err : {%v}", targetNamePath, err)
+		return HostIDs, err
+	}
+
+	logger.Debugf("targetname files matches were found : {%v}", matches)
+
+	for _, targetPath := range matches {
+		logger.Debugf("Check if targetfile (%s) value is Online.", targetPath)
+		targetName, err := o.executer.IoutilReadFile(targetPath)
+		if err != nil {
+			logger.Warningf("Could not read target name from file : {%v}, error : {%v}", targetPath, err)
+			continue
+		}
+
+		if strings.EqualFold(strings.TrimSpace(string(targetName)), "online") {
+			re := regexp.MustCompile("host([0-9]+)")
```

Comment:
1. move it out of the for-loop
2. I think "/host([0-9]+)/" is more accurate, is it?

Reply:
Done